### PR TITLE
Migration to ViewPager 2

### PR DESCRIPTION
**Pull request by vbuberen**
## :page_facing_up: Context
Google encourages developers to try the new ViewPager 2, since it offers quite a lot optimizations.
More info on changes: https://medium.com/google-developer-experts/exploring-the-view-pager-2-86dbce06ff71
The library is in `release candidate` state, so I am sure that no changes should appear there and we can safely update Chucker to use latest components. 
Additionally, this migration allowed to make the code cleaner and more concise.

## :pencil: Changes
Added dependency to corresponding library.
Updated ViewPager adapters in `MainActivity` and `TransactionActivity` to be able to work with ViewPager2 as well as changed layouts to use `ViewPager2`.

### CODE REVIEW

File: library/src/main/java/com/chuckerteam/chucker/internal/ui/MainActivity.kt [hunk 1]
```diff
@@ -74,13 +61,38 @@ class MainActivity :
...
+                super.onPageSelected(position)
+                if (position == HomePageAdapter.NETWORK_SCREEN_POSITION) {
+                    Chucker.dismissTransactionsNotification(this@MainActivity)
+                } else {
+                    Chucker.dismissErrorsNotification(this@MainActivity)
+                }
+            }
+        })
+
+        val tabLayout = findViewById<TabLayout>(R.id.tabLayoutHome)
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/ui/MainActivity.kt]:
```suggestion
        val tabLayout : TabLayout = findViewById(R.id.tabLayoutHome)
```

File: library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt [hunk 2]
```diff
@@ -103,16 +98,30 @@ internal class TransactionActivity : BaseChuckerActivity() {
...
 
-    private fun setupViewPager(viewPager: ViewPager) {
-        adapter = TransactionPagerAdapter(this, supportFragmentManager)
-        viewPager.adapter = adapter
-        viewPager.addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
-            override fun onPageSelected(position: Int) {
-                selectedTabPosition = position
-                populateUI(transaction)
+    private fun setupViewPager() {
+        val viewPager = findViewById<ViewPager2>(R.id.viewPagerTransaction)
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt]:
```suggestion
        val viewPager : ViewPager2 = findViewById(R.id.viewPagerTransaction)
```

File: library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt [hunk 3]
```diff
@@ -103,16 +98,30 @@ internal class TransactionActivity : BaseChuckerActivity() {
...
+            registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
+                override fun onPageSelected(position: Int) {
+                    selectedTabPosition = position
+                    populateUI(transaction)
+                }
+            })
+            currentItem = selectedTabPosition
+        }
+
+        val tabLayout = findViewById<TabLayout>(R.id.tabLayoutTransaction)
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt]:
```suggestion
        val tabLayout : TabLayout = findViewById(R.id.tabLayoutTransaction)
```

Comment (vbuberen) [reply library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt]: I think that we should drop all these `findViewById` usages. But in another PR later.

**Comment by vbuberen** [comment]
1. Don't see leaks. After merging latest `develop` changes with `ViewModel` see no leaks too.
2. Yeah, strange order of menu creations in fragments. Interesting.

**Comment by cortinico** [comment]
> * Don't see leaks. After merging latest `develop` changes with `ViewModel` see no leaks too.

Same here, so you can drop this now.

> * Yeah, strange order of menu creations in fragments. Interesting.

This is definitely a regression. Are you able to have a look at it before we merge this?

File: library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt (hunk 3 above)
Comment (cortinico) [reply library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionActivity.kt]: Yup, feel free to drop it for now.

**Comment by vbuberen** [comment]
Yes, I definitely want to investigate it. Tried a few things yesterday, but wasn't able to resolve just yet. 
It is something with transactions list menu invalidation before errors list menu starts its initialization.
Also, it happens only during first switch between views.

**Comment by vbuberen** [comment]
Tried to debug and find the problem, but had no luck so far.
Somehow `onCreateOptionsMenu()` is called in `TransactionListFragment` again before calling `onCreateOptionsMenu()` in `ErrorListFragment`
Opened an issue on Google's issue tracker: https://issuetracker.google.com/issues/147550812

**Comment by cortinico** [comment]
> Tried to debug and find the problem, but had no luck so far.

What do we do here? Do we want to move it to `3.2.0` or we include it as it is? @vbuberen

**Comment by vbuberen** [comment]
I would postpone to 3.2.0 to not compromise on quality. Ideally, we should strive to have stable versions as bugless as possible (even withoutsuch minor issues), so users could be confident in library quality.

**Comment by vbuberen** [comment]
Since no updates to ViewPager2 were presented moving further to versions after `3.2.0`

**Comment by vbuberen** [comment]
Yesterday Google released 1.1.0-alpha01 and it has fix for the issue with options menu. Checked it and everything works fine. Going to wait till it gets stable and we will be able to finally merge this PR.

**Comment by cortinico** [comment]
> Yesterday Google released 1.1.0-alpha01 and it has fix for the issue with options menu

Great news, let's wait for the stable and we can proceed them. Should we have a label like `on hold` for PRs like this one?

**Comment by vbuberen** [comment]
I think `do not merge` works fine here. Don't want to make things too complicated.

**Comment by vbuberen** [comment]
Another test
/rebase

**Comment by cortinico** [comment]
> Another test
> /rebase

I'm afraid you can't rebase this branch as it has a non linear git history (there are a lot of `Merge latest develop changes`). It works fine if there are no merge commits in the branch to rebase.

**Comment by vbuberen** [comment]
Yeah, I see.

**Comment by vbuberen** [comment]
It's been a year since I opened this PR and it looks like we still can't merge it.
Today I tried to update the PR with latest changes from `develop` branch, which I believed would unblock us from merging. 

Since we have no `ViewPager` in the `MainActivity` anymore, we should have no issues with options menu items there, which were one of the reasons why the PR is still unmerged. However, while testing things again I noticed another issue:

Sometimes wrong set of option menu items appears in the toolbar. For example here is a screenshot of `Overview` tab, where search and save should be hidden: 
menu_items

Due to this issue and the fact that Google hasn't released any updates to ViewPager2 1.0.0 except some alpha (which we can't use to not create potential issues like #469) I am convinced that we are fine with older ViewPager, which has no issues for our usage and needs no workarounds.

Closing this PR. Hope that in future we could return to this when `ViewPager2` gets more love from Google.

**Pull request closed without merging**